### PR TITLE
Update MMR Runtime API - add proof batching

### PR DESCRIPTION
**Pull request by Wizdave97**
Companion PR to https://github.com/paritytech/substrate/pull/10635

### CODE REVIEW

File: runtime/kusama/src/lib.rs [hunk 1]
```diff
@@ -1797,10 +1797,31 @@ sp_api::impl_runtime_apis! {
 		}
 
 		fn mmr_root() -> Result<Hash, mmr::Error> {
+			// dummy implementation due to lack of MMR pallet.
+			Err(mmr::Error::GenerateProof)
+		}
+		fn generate_batch_proof(_leaf_indices: Vec<u64>)
```
Comment (acatangiu) [inline runtime/kusama/src/lib.rs:1803]:
```suggestion
		fn mmr_root() -> Result<Hash, mmr::Error> {
			// dummy implementation due to lack of MMR pallet.
			Err(mmr::Error::GenerateProof)
		}
		
		fn generate_batch_proof(_leaf_indices: Vec<u64>)
```

File: runtime/kusama/src/lib.rs [hunk 2]
```diff
@@ -1797,10 +1797,31 @@ sp_api::impl_runtime_apis! {
 		}
 
 		fn mmr_root() -> Result<Hash, mmr::Error> {
+			// dummy implementation due to lack of MMR pallet.
+			Err(mmr::Error::GenerateProof)
```
Comment (acatangiu) [inline runtime/kusama/src/lib.rs:1801]:
```suggestion
			Err(mmr::Error::GetRoot)
```

File: runtime/polkadot/src/lib.rs [hunk 3]
```diff
@@ -1945,6 +1945,29 @@ sp_api::impl_runtime_apis! {
 		}
 
 		fn mmr_root() -> Result<Hash, mmr::Error> {
+			// dummy implementation due to lack of MMR pallet.
+			Err(mmr::Error::GenerateProof)
```
Comment (acatangiu) [inline runtime/polkadot/src/lib.rs:1949]:
```suggestion
			Err(mmr::Error::GetRoot)
```

File: runtime/test-runtime/src/lib.rs [hunk 4]
```diff
@@ -932,10 +932,32 @@ sp_api::impl_runtime_apis! {
 		}
 
 		fn mmr_root() -> Result<Hash, mmr::Error> {
+			// dummy implementation due to lack of MMR pallet.
+			Err(mmr::Error::GenerateProof)
```
Comment (acatangiu) [inline runtime/test-runtime/src/lib.rs:936]:
```suggestion
			Err(mmr::Error::GetRoot)
```

File: runtime/westend/src/lib.rs [hunk 5]
```diff
@@ -1351,6 +1351,29 @@ sp_api::impl_runtime_apis! {
 		}
 
 		fn mmr_root() -> Result<Hash, mmr::Error> {
+			// dummy implementation due to lack of MMR pallet.
+			Err(mmr::Error::GenerateProof)
```
Comment (acatangiu) [inline runtime/westend/src/lib.rs:1355]:
```suggestion
			Err(mmr::Error::GetRoot)
```

**Review by acatangiu — APPROVED** [review]

**Review by svyatonik — APPROVED** [review]

**Comment by drahnr** [comment]
Is there a particular reason we are not creating a dedicated error value named `PalletNotIncluded` rather than adding comments?

**Review by andresilva — APPROVED** [review]

**Comment by drahnr** [comment]
Could you elaborate the rationale behind implementing the mmr API which only errors rather than omitting the implementation entirely? Where does the requirement having that trait in the runtime stem from?

**Comment by acatangiu** [comment]
> Could you elaborate the rationale behind implementing the mmr API which only errors rather than omitting the implementation entirely? Where does the requirement having that trait in the runtime stem from, for those where it'd be error only?

@andresilva or @tomusdrw might have the full context, but I believe BEEFY+MMR were previously deployed on all chains, then decided to be disabled and reworked, but the APIs kept in the runtimes for compatibility maybe?

I believe the thinking was they'd be relatively quickly re-enabled.

Right now, the timeline for BEEFY+MMR to reach Polkadot is end of this year with Kusama sooner.

If we remove the stubs now, and re-add them as fully functional APIs later would we need runtime upgrades? Not sure how big the hassle is and if it's worth it...

**Comment by Wizdave97** [comment]
@acatangiu  is there anything still holding stalling this PR?

File: runtime/kusama/src/lib.rs [hunk 6]
```diff
@@ -1807,14 +1807,36 @@ sp_api::impl_runtime_apis! {
...
 		fn mmr_root() -> Result<Hash, mmr::Error> {
 			// dummy implementation due to lack of MMR pallet.
-			Err(mmr::Error::Verify)
+			Err(mmr::Error::PalletNotIncluded)
 		}
 
+		fn generate_batch_proof(_leaf_indices: Vec<u64>)
+			-> Result<(Vec<mmr::EncodableOpaqueLeaf>, mmr::BatchProof<Hash>), mmr::Error>
+		{
+			// dummy implementation due to lack of MMR pallet.
```
Comment (drahnr) [inline runtime/kusama/src/lib.rs:1821]: nit: This comment should be a doc comment on the error variant.

File: bridges/bin/millau/runtime/src/lib.rs [hunk 7]
```diff
@@ -681,35 +683,59 @@ impl_runtime_apis! {
...
 
 		fn verify_proof(leaf: EncodableOpaqueLeaf, proof: MmrProof<MmrHash>)
 			-> Result<(), MmrError>
 		{
-			pub type Leaf = <
-				<Runtime as pallet_mmr::Config>::LeafData as LeafDataProvider
-			>::LeafData;
 
-			let leaf: Leaf = leaf
+			pub type MmrLeaf = <<Runtime as pallet_mmr::Config>::LeafData as LeafDataProvider>::LeafData;
```
Comment (drahnr) [inline bridges/bin/millau/runtime/src/lib.rs:697]: Does this type have to be `pub`?

**Review by drahnr — APPROVED** [review]
LGTM, besides the few minor nits mentioned

**Comment by acatangiu** [comment]
@Wizdave97 please pull in latest `polkadot/master`

**Comment by acatangiu** [comment]
@paritytech/polkadot-review runtime files are touched with trivial change, but 2 approvals from you are required nevertheless.

**Comment by Wizdave97** [comment]
> @paritytech/polkadot-review runtime files are touched with trivial change, but 2 approvals from you are required nevertheless.

Merged the latest changes

**Comment by drahnr** [comment]
CI still stumbles upon some of the changes (did not dig into it yet). I think most reviewers prefer all green PRs before doing a detailed review.

**Comment by acatangiu** [comment]
> CI still stumbles upon some of the changes (did not dig into it yet). I'll bug more reviewers once that's achieved.

Makes sense.

Right now, the PR is as green as it can get for a companion.
The substrate PR, for which this is a companion, is 100% green: https://github.com/paritytech/substrate/pull/10635

**Comment by acatangiu** [comment]
The required approves set is kinda tight:

```
usersToAskForReview Map(4) {
  'gavofyork' => { teams: Set(1) { 'locks-review' } },
  'shawntabrizi' => { teams: Set(1) { 'locks-review' } },
  'bkchr' => { teams: Set(1) { 'locks-review' } },
  'rphmeier' => { teams: Set(1) { 'locks-review' } }
}
```

@paritytech/locks-review please take a look, changes are trivial at the runtime level.

**Review by bkchr — APPROVED** [review]

**Comment by ordian** [comment]
bot merge

**Pull request merged**